### ISSUE TITLE
Autotype doesn't support some KeePass codes

### ISSUE THREAD

**Issue by TheCycoONE** [keepassxreboot/keepassxc#977]
Symbols supported in Keepass autotype, e.g. + for shift, are not supported in keepassxc resulting in bad autotype behaviour for imported or shared databases.

## Expected Behavior
The autotype sequence I use in KeyPass is: `+{END}{UserName}{TAB}+{END}{Password}{ENTER}` which ensures that the username and password replace whatever is already entered in the field. (shift + end)<username>(tab)(shift + end)<password>(enter).

Expected behaviour would be to match the list from http://keepass.info/help/base/autotype.html

## Current Behavior
In KeePassXC I'm instead seeing the + prepended on my username and password.

## Context
I share the same database (through frequency syncing) across a number of devices, and operating systems. On Windows I'm running KeyPass 2, and on Linux and OSX I'm using KeyPassXC. Some of the sites I log in to automatically populate the username and password fields, so I added the +{END} sequence to my autotype default to address those pages.

## Debug Info
KeePassXC - Version 2.2.0
Revision: caa49a8ef3ee28ed478192389b21d61107b3b8e0

Libraries:
- Qt 5.9.0
- libgcrypt 1.7.7

Operating system: macOS Sierra (10.12)
CPU architecture: x86_64
Kernel: darwin 16.7.0

Enabled extensions:
- KeePassHTTP
- Auto-Type
- YubiKey

**Comment by TheCycoONE** [comment]
It seems for my use case, the solution would be to use {CLEARFIELD} which is supported by both.

**Comment by TheZ3ro** [comment]
Yes, you should use `{CLEARFIELD}` when possible. We don't have implemented the `+` special character right now.

**Comment by droidmonkey** [comment]
The plus special character seems like a mistake to me. Why have it different then the given standard of wrapping commands with { }

**Comment by TheZ3ro** [comment]
@droidmonkey I don't know but plain plus `+` is equal to the <kbd>SHIFT</kdb> button, if you want to print the plus symbol you should use `{+}` according to KeePass reference.

I will implement such symbols for 2.3.0 version to keep our autotype compliant to those ref.

**Comment by droidmonkey** [comment]
That is so backwards it makes my head hurt. {+} should be <kbd>SHIFT</kbd> to match the rest of the convention!

**Comment by TheCycoONE** [comment]
@droidmonkey If it helps your conceptual model, the three replacements without {..} are modifiers not substitutions: `+` for `shift`, `^` for `control` and `%` for alt.

**Comment by wofwofwof** [comment]
Anybody working on this? I really miss the option to use ALT and and CTRL. I use this keys to switch between my virtual screens. 

I've just looked into the source. This change would require to refactor AutoTypeChar and then all AutoTypePlatformPlugins. There seems to be lots of work going on in the AutoType code so I hesitate to refactor it. Anybody planning to work here in near future?

cheers
wof

**Comment by droidmonkey** [comment]
You're welcome to fix this if you like.

**Comment by jwenz723** [comment]
I'm trying to understand what I would need to do to add the 'shift' functionality. Wondering if someone can tell me why this doesn't work:

https://github.com/keepassxreboot/keepassxc/blob/010d44eab7a84e457b6b72ef90c0f32d69511084/src/autotype/mac/AutoTypeMac.cpp#L318-L319

**Comment by droidmonkey** [comment]
Because shift is a key modifier. It is sent to the OS in addition to any other keys pressed. Think of it this way, pressing shift all by itself does absolutely nothing. You have to add another key to make it work.

**Comment by Simplenuity** [comment]
I've set my default Auto-Type sequence to
`{CLEARFIELD}{USERNAME}{TAB}{CLEARFIELD}{PASSWORD}`
and it used to work as expected. 
I just realized that it meanwhile doesn't work anymore, i.e. both fields don't get cleared anymore.
Not sure if it's related to upgrading to 2.4.1 via ppa some time ago.

**Comment by radicarl** [comment]
Hello,
I need this too. In my case I have a form with
username
pin+token
password

The token is generated by a token generator and must be entered manual. My autotype should enter username, pin, password and the tab back (shift+tab) to the pin+token field. Without a working shift-modifier I have to select the field by myself

**Comment by OLLI-S** [comment]
I also recognized a missing Auto-Type code: **{C:Comment}**

In the documentation of KeePass Auto-Type (https://keepass.info/help/base/autotype.html) this  code is not listed, but Dominik Reichl sent me the information by mail when I suggested a new feature.

The problem is that some websites (like Amazon or PayPal) have two different login masks:
- one where the username and the password is on one page
- one where the username and the password are on two different pages 

Amazon and PayPal display still use these two different types (depending where you come from when you access these sites).

So in KeePass I have for Amazon and PayPal two different Auto-Type sequences.
When I am at the Amazon website and press the Auto-Type hotkey, then in the "Auto-Type Entry Selection" dialog of KeePass I can not see the full Auto-Type sequence (to differentiate the two sequences).

KeePass allows to enter a comment in the Auto-Type sequence:
`{USERNAME}{TAB}{PASSWORD}{ENTER}{C:Felder auf einer Seite}`
`{USERNAME}{TAB}{ENTER}{DELAY 2000}{PASSWORD}{ENTER}{C:Felder auf mehreren Seiten}`
The comments mean "Fields on one page" and "Fields on multiple pages"

In the "Auto-Type Entry Selection" dialog the comments are shown when you have set to see the column "Sequence - Comments":

![image](https://user-images.githubusercontent.com/8037790/72206744-ce0bbf00-3491-11ea-85f3-9e9659e59d48.png)

## Suggestion
Currently you show the following message when the syntax of the Auto-Type statement is not correct:

![image](https://user-images.githubusercontent.com/8037790/72206809-6c982000-3492-11ea-8c09-20a2b0b6828f.png)

I suggest that you show the invalid part in this message, like:
```
The following statement in the Auto-Type statement is not correct or not supported:
{C:Felder auf mehreren Seiten}
```
This way users see what part is not correct/not supported.

I think it is not necessary to add a comment in Auto-Type (like in KeePass).
Otherwise I can add a issue for this suggestion.

**Comment by sergeevabc** [comment]
Agggrrrhhh! O ```+{TAB}```, where art thou?

**Comment by rzumer** [comment]
Is `T-REPLACE-RX` unsupported? Unfortunately I will have to go back to the official client if not, as I have a complex auto-type sequence for an application that has a non-standard input method.

**Comment by droidmonkey** [comment]
Is that even a documented thing?

**Comment by rzumer** [comment]
It's been for years. https://keepass.info/help/base/placeholders.html

**Comment by droidmonkey** [comment]
That looks unnecessarily complicated and will never be implemented.

**Comment by rzumer** [comment]
Thanks for addressing it, I suppose I will have to stick with the official Keepass client that supports auto-type in unnecessarily complicated applications.

**Comment by droidmonkey** [comment]
Can you explain your use case?

**Comment by rzumer** [comment]
Like I said, it is for auto-type in an application that uses a non-standard input method. Without string replacement, I would have to mangle my username in order for it to be written correctly. I'm surprised that clients would choose not to implement part of the Keepass "specification" and forego compatibility with auto-type sequences written in the official client, but I respect the decision. For me this makes the official client more attractive.

**Comment by droidmonkey** [comment]
You should just put the mangled username in the advanced attributes along with your group name. Your sequence can simply reference those attributes using the `{S:<attrname>}` syntax.

**Comment by rzumer** [comment]
That does not work because the custom field is entered as-is without considering modifiers, so the actual key combination that needs to be typed is impossible to represent as far as I know. Assuming that works as designed, it is impossible to write the "mangled username" without text replacement from the auto-type sequence.

**Comment by ljhmily** [comment]
> In the "Auto-Type Entry Selection" dialog the comments are shown when you have set to see the column "Sequence - Comments":
how does this "auto-type entry selection" UI get?
